Optimize revision recording, speed up update dashboard API by 25% 🚀

qnkhuat opened the pull request:
An attempt to resolve https://github.com/metabase/metabase/issues/34721 by reducing the number of DB calls when recording revision.

Other changes:
- adding schema for event system in "metabase.events.schema"
- decouple the info in the `event` of `(events/publish-event! topic event)`, before it's used to store everything it's needed for the event to process, it's usually a DB object like Card/Dashboard, and assoced with `actor-id`, `dashcards`... changing this to `{:object dashboard, :other-info 1}` to makes it explicit what's needed. also we can reuse the `object` to record revision so we don't have to refetch it.

Benchmarking ([script](https://gist.github.com/qnkhuat/65d63319fa3d63e52b08c1afe17094f4)):

master
- `GET /api/dashboard/:id` reports 16 DB calls
  -  takes ~14.14ms (N=1000)
- `POST /api/dashboard` reports 23 DB calls
  - takes ~14.7ms (N= 1000)
- `PUT /api/dashboard/:id` reports 22 DB calls
  - takes ~16.19ms (N=1000)
- `PUT /api/dashboard/cards` report 50 DB calls
  - takes ~ 34.6ms (N=1000)

this branch
- `GET /api/dashboard/:id` reports 16 DB calls
  -  takes ~12ms (N=1000)
- `POST /api/dashboard/:id` reports 20 DB calls
  - takes ~11.4ms (N= 1000)
- `PUT /api/dashboard/:id` reports 25 DB calls
  - takes ~14.2ms (N=1000)
- `PUT /api/dashboard/cards` report 39 DB calls
  - takes ~ 25.4ms (N=1000)

qnkhuat committed:
Schema for revision APIs

qnkhuat committed:
batched hydrating the series

qnkhuat committed:
Makes sure we hydrate series if needed in serialize-dashboard

qnkhuat committed:
Adding schema for events and starting with dashboard events

qnkhuat committed:
remove debug code

qnkhuat committed:
schema for the whole thing and fix tests

qnkhuat committed:
hoping we can drop activity alltogether

qnkhuat committed:
Merge branch 'master' into ngoc-optimize-revision

qnkhuat committed:
remove object->user-id and object->model-id

qnkhuat committed:
still select but more efficient

qnkhuat committed:
test fixes

qnkhuat committed:
test fixes 😎

qnkhuat commented on src/metabase/api/card.clj:
```diff
@@ -608,20 +608,6 @@ saved later when it is ready."
     (validation/check-embedding-enabled)
     (api/check-superuser)))
 
-(defn- publish-card-update!
```
there are no subscribers for card-unarchive and card-archive, also it has never been published before because archived and `(:card archive)` is identical when it's called this function.

so I've removed it

qnkhuat commented on src/metabase/events/revision.clj:
```diff
@@ -12,25 +10,28 @@
+       (revision/push-revision! {:entity       model
                                  :id           id
-                                 :object       (api/check-404 (t2/select-one model :id id))
+                                 :object       object
```
optimization: no need to select the object again

qnkhuat commented on src/metabase/models/dashboard.clj:
```diff
@@ -216,11 +216,19 @@
-                           (-> (apply dissoc dashboard-card excluded-columns-for-dashcard-revision)
-                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))
-      (assoc :tabs (map #(apply dissoc % excluded-columns-for-dashboard-tab-revision) (tabs dashboard)))))
+  (let [cards (or (:cards dashboard)
```
optimization: only hydrates when needed

qnkhuat replied:
renamed to dashcards https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -89,16 +88,23 @@
 ;;; --------------------------------------------------- HYDRATION ----------------------------------------------------
 
-(mi/define-simple-hydration-method series
+(mi/define-batched-hydration-method series
```
series has always been batch hydrated, so making it more efficient

qnkhuat commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -111,8 +117,9 @@
 (mu/defn retrieve-dashboard-card
```
this is currently only used in test tho.

qnkhuat commented on src/metabase/events/activity_feed.clj:
```diff
@@ -47,99 +50,120 @@
-          ;; plus a `:dashcards` attribute which is a vector of the cards added/removed
-          (-> (t2/select-one [Dashboard :description :name], :id (events/object->model-id topic obj))
-              (assoc :dashcards (for [{:keys [id card_id]} dashcards]
-                                  (-> (t2/select-one [Card :name :description], :id card_id)
```
N+1 with the number of dash cards 🙀

calherries commented on src/metabase/events/activity_feed.clj:
```diff
@@ -47,99 +50,120 @@
-  [topic object]
-  (let [details-fn  #(select-keys % [:name :description :revision_message])
-        table-id    (:table_id object)
+  [topic event]
```
Again, I think it would be nicer to destructure all the keys of event here

qnkhuat replied:
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

calherries commented on src/metabase/models/revision.clj:
```diff
@@ -153,37 +154,37 @@
-  [model id]
+  [model :- [:fn mdb.u/toucan-model?]
+   id    :- pos-int?]
   {:pre [(mdb.u/toucan-model? model) (integer? id)]}
```
You don't need to duplicate the schema with the `:pre` block
```suggestion
```

calherries reviewed (approved):
Approving because it works and my suggestions are all stylistic. I do want `:actor-id` renamed to `:user-id` everywhere though.

My biggest suggestion would be to make smaller PRs if you can help it next time. There's far too much stuff happening in this PR to review quickly and well.

calherries commented on src/metabase/events/schema.clj:
```diff
@@ -0,0 +1,133 @@
+(ns metabase.events.schema
```
For a follow-up PR: Rather than centralize event schemas in one place and have all this code to merge them together, I suggest dropping the schemas and thinking about how to create a multimethod like `validate-event` that is dispatched by event topic. 

Imagine if all our functions has schemas in one namespace! We'd waste a lot of our dev time just switching between files and following references

qnkhuat replied:
see my other [comment](https://github.com/metabase/metabase/pull/34947/files#r1375607002) for details on why I chose this approach.

calherries replied:
Still, it feels desirable to change `events.schema/topic->schema` to be a multimethod so you don't have to merge all the schemas into a map and you can define them in any namespace.

calherries commented on src/metabase/events/schema.clj:
```diff
@@ -0,0 +1,133 @@
+
+;; dashboard events
+
+(def ^:private dashboard-default
+  (mc/schema
+   [:map {:closed true}
+    [:actor-id pos-int?]
+    [:object [:fn #(t2/instance-of? :model/Dashboard %)]]]))
+
+(def ^:private dashboard-update-with-dashcards
+  (mc/schema
+   [:merge
+    dashboard-default
+    [:map {:closed true}
+     [:dashcards [:sequential [:map [:id pos-int?]]]]]]))
+
+(def ^:private dashboard-update-with-tab-ids
+  (mc/schema
+   [:merge
+    dashboard-default
+    [:map {:closed true}
+     [:tab-ids [:sequential pos-int?]]]]))
+
+(def ^:private dashboard-events-schemas
+  {:event/dashboard-read             dashboard-default
+   :event/dashboard-create           dashboard-default
+   :event/dashboard-update           dashboard-default
+   :event/dashboard-reposition-cards dashboard-update-with-dashcards
+   :event/dashboard-remove-cards     dashboard-update-with-dashcards
+   :event/dashboard-add-cards        dashboard-update-with-dashcards
+   :event/dashboard-add-tabs         dashboard-update-with-tab-ids
+   :event/dashboard-update-tabs      dashboard-update-with-tab-ids
+   :event/dashboard-remove-tabs      dashboard-update-with-tab-ids})
```
This could be much briefer
```clojure
(let [default-schema [:map {:closed true}
                      [:actor-id pos-int?]
                      [:object [:fn #(t2/instance-of? :model/Dashboard %)]]]
      with-dashcards (mut/assoc default-schema :dashcards [:sequential [:map [:id pos-int?]]])
      with-tabs      (mut/assoc default-schema :tab-ids [:sequential pos-int?])]
  (def ^:private dashboard-events-schemas
    {:event/dashboard-read             dashboard-default
     :event/dashboard-create           dashboard-default
     :event/dashboard-update           dashboard-default
     :event/dashboard-reposition-cards dashboard-update-with-dashcards
     :event/dashboard-remove-cards     dashboard-update-with-dashcards
     :event/dashboard-add-cards        dashboard-update-with-dashcards
     :event/dashboard-add-tabs         dashboard-update-with-tab-ids
     :event/dashboard-update-tabs      dashboard-update-with-tab-ids
     :event/dashboard-remove-tabs      dashboard-update-with-tab-ids}))
```

qnkhuat replied:
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat committed:
Merge branch 'master' into ngoc-optimize-revision

qnkhuat commented on src/metabase/events.clj:
```diff
@@ -107,41 +108,32 @@
+     (assert (map? event)
+             (format "Invalid event %s: event must be a map." (pr-str event)))
+     (try
+      (if-let [schema (events.schema/topic->schema topic)]
+        (mu/validate-throw schema event)
+        (log/warn (format "Missing schema for topic: %s" topic)))
```
on schema living with the code: the thing with the event system is that one topic key can have multiple handlers, 

take `:event/card-create` for example, there are 4 subscribers defined in 4 namespaces.

So defining a schema on defmethod will force us to duplicate the schema everywhere.

but that is not the biggest problem tho, I think we can have a single ns for schema and reuse it across event namespaces. The biggest problem right now is methodical haven't supported it https://github.com/camsaul/methodical/issues/145

qnkhuat replied:
> Also I wouldn't warn if a schema is missing for a topic

yea I wanted it to warn to aid me in adding events with missing schema. now it's done so I'll remove the log.

qnkhuat replied:
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on src/metabase/models/pulse.clj:
```diff
@@ -507,9 +506,7 @@
-  (let [pulse-id (create-notification-and-add-cards-and-channels! kvs cards channels)]
-    ;; return the full Pulse (and record our create event)
-    (events/publish-event! :event/pulse-create (retrieve-pulse pulse-id))))
+  (retrieve-pulse (create-notification-and-add-cards-and-channels! kvs cards channels)))
```
I've moved the publish-event into the api namespace.

qnkhuat committed:
- remove log/warn when event schema not found
    - destructure the event at arg definition
    - shorten events chema using let block
    - toucan2 is smart enough to batched hydrated with a single object

qnkhuat committed:
fix record-activity

qnkhuat commented on src/metabase/events/activity_feed.clj:
```diff
@@ -21,24 +20,28 @@
-                             (log/error e (tru "Error preprocessing query:")))))
-        database-id (some-> query :database u/the-id)
-        table-id    (mbql.u/query->source-table-id query)]
+  [topic event]
```
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on src/metabase/events/activity_feed.clj:
```diff
@@ -47,99 +50,120 @@
+                  :event/dashboard-create       (create-delete-details event)
+                  :event/dashboard-delete       (create-delete-details event)
+                  :event/dashboard-add-cards    (add-remove-card-details event)
+                  :event/dashboard-remove-cards (add-remove-card-details event))})))
```
cool https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on src/metabase/events/driver_notifications.clj:
```diff
@@ -15,10 +15,11 @@
-    (catch Throwable e
-      (log/warnf e "Failed to process driver notifications event. %s" topic))))
+   ;; notify the appropriate driver about the updated database
+   (let [database (:object event)]
```
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on src/metabase/models/revision.clj:
```diff
@@ -153,37 +154,37 @@
-         (t2/exists? entity :id id)
-         (map? object)]}
-  (let [serialized-object (serialize-instance entity id (dissoc object :message))
+  [info :- [:map {:closed true}
```
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on src/metabase/models/dashboard_card.clj:
```diff
@@ -111,8 +117,9 @@
-      (t2/hydrate :series)))
+  (-> (t2/select :model/DashboardCard :id id)
+      (t2/hydrate :series)
+      first))
```
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat commented on test/metabase/api/dashboard_test.clj:
```diff
@@ -1445,7 +1445,7 @@
 (defn- dashcard-like-response
   [id]
-  (t2/hydrate (t2/select-one DashboardCard :id id) :series))
+  (first (t2/hydrate (t2/select DashboardCard :id id) :series)))
```
https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

qnkhuat committed:
actor-id -> user-id

qnkhuat commented on src/metabase/events/revision.clj:
```diff
@@ -12,25 +10,28 @@
+     (let [id      (or (:id object)
+                       (throw (ex-info "Event does not have ID associated with it"
+                                       {:mode model :event event})))
+           user-id (or (:actor-id event)
```
did this in 0fa9b07033

qnkhuat committed:
add missing schemas and removed some unused ones

qnkhuat committed:
fix schema for card-query

calherries commented on src/metabase/query_processor/middleware/process_userland_query.clj:
```diff
@@ -92,12 +92,12 @@
-                                                   :context      (:context execution-info)
-                                                   :ignore_cache (get-in execution-info
-                                                                         [:json_query :middleware :ignore-cached-results?])}))
+         (events/publish-event! :event/card-query {:user-id (:executor_id execution-info)
+                                                   :object  {:id           (:card_id execution-info)
+                                                             :cached       (:cached acc)
+                                                             :context      (:context execution-info)
+                                                             :ignore_cache (get-in execution-info
+                                                                                   [:json_query :middleware :ignore-cached-results?])}}))
```
Why did you move this data under an `:object` key? It doesn't look like an object to me, just a bunch of data

qnkhuat replied:
https://github.com/metabase/metabase/blob/7eb27d08d973ba987657d41316e8a48b6090a75b/src/metabase/events/view_log.clj#L81

this is the code that consumes this event and it's currently used to process `:event/dashboard-read`, `:event/table-read`, `:event/card-read` as well.

so I make it an object in order to not have to modify this defmethod.

another option is to create a new defmethod just to handle `:event/card-query`, is that better?

calherries replied:
Yes, I think creating a new defmethod to handle `:event/card-query` is a better option. We should avoid reusing names of keys inconsistently. The current handler for `:metabase.events.view-log/event` is effectively handling two different formats of data - one for actual views and one for queries.

qnkhuat committed:
rework the event/card-read event

calherries reviewed (approved):
Looks good, thanks

qnkhuat committed:
fix kondo

qnkhuat committed:
Merge branch 'master' into ngoc-optimize-revision

qnkhuat committed:
fix revision str test

qnkhuat commented on src/metabase/models/revision/diff.clj:
```diff
@@ -119,14 +119,13 @@
-  (clojure.core.match/match [:name x y]
-    [:name _ _]
-    "Ok"))
+      (loop [ks               ks
```
we had a failing e2e test because of this https://github.com/metabase/metabase/actions/runs/6714770177/job/18249901919?pr=34947

qnkhuat committed:
fix the test

qnkhuat committed:
Merge branch 'master' into ngoc-optimize-revision
